hugr-core/src/extension/op_def.rs

```diff
@@ -142,26 +139,80 @@ impl CustomValidator {
     ) -> Self {
         Self {
             poly_func: poly_func.into(),
-            validate: Box::new(validate),
+            validate: Some(Box::new(validate)),
         }
     }
 }
 
 /// The two ways in which an OpDef may compute the Signature of each operation node.
-#[derive(serde::Deserialize, serde::Serialize)]
 pub enum SignatureFunc {
     // Note: except for serialization, we could have type schemes just implement the same
     // CustomSignatureFunc trait too, and replace this enum with Box<dyn CustomSignatureFunc>.
     // However instead we treat all CustomFunc's as non-serializable.
     /// A PolyFuncType (polymorphic function type), with optional custom
     /// validation for provided type arguments,
-    #[serde(rename = "signature")]
     PolyFuncType(CustomValidator),
-    #[serde(skip)]
     /// A custom binary which computes a polymorphic function type given values
     /// for its static type parameters.
     CustomFunc(Box<dyn CustomSignatureFunc>),
+    /// Declaration specified a custom binary but it was not provided.
+    MissingComputeFunc,
+}
+
+mod serialize_signature_func {
+    use serde::{Deserialize, Serialize};
+
+    use super::{PolyFuncTypeRV, SignatureFunc};
+    #[derive(serde::Deserialize, serde::Serialize)]
+    struct SerSignatureFunc {
+        /// If the type scheme is available explicitly, store it.
+        signature: Option<PolyFuncTypeRV>,
+        /// Whether an associated binary function is expected.
+        /// If `signature` is `None`, a true value here indicates a custom compute function.
+        /// If `signature` is not `None`, a true value here indicates a custom validation function.
+        binary: bool,
+    }
+
+    pub(super) fn serialize<S>(
+        value: &super::SignatureFunc,
+        serializer: S,
+    ) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match value {
+            SignatureFunc::PolyFuncType(custom) => SerSignatureFunc {
+                signature: Some(custom.poly_func.clone()),
+                binary: custom.validate.is_some(),
+            },
+            SignatureFunc::CustomFunc(_) => SerSignatureFunc {
+                signature: None,
+                binary: true,
+            },
+            SignatureFunc::MissingComputeFunc => SerSignatureFunc {
+                signature: None,
+                binary: false,
+            },
+        }
+        .serialize(serializer)
+    }
+
+    pub(super) fn deserialize<'de, D>(deserializer: D) -> Result<super::SignatureFunc, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let SerSignatureFunc { signature, .. } = SerSignatureFunc::deserialize(deserializer)?;
+
+        // TODO for now, an indication of expected custom validation function is ignored.
+        Ok(if let Some(sig) = signature {
```

Comment:
nit - this is something like `signature.map_or(MissingComputeFunc, super::SignatureFunc::From)` where the `MissingComputeFunc` is the default for `None`